Relax matching criteria and patch question-related issues

c0j0s opened the pull request:
# Pull Request

## Description
The matching logic now supports matching by common preference in `complexity` and `topic`. 
Upon matched, the success view should display their common preferences that will be used to query for questions.
After the owner requests to start the collab, the lobby will query for questions, and  redirect clients to collab page once the question is ready,

## Related Issue(s)
Close #91 

## Changes Made
- Relax matching criteria
- Support question selection for collab setup
- Add toast messages for the question page
- Update question wrapper layer
- Consume `getTopics` in matching card and question modal.
- Fix issue with hidden buttons in some matching lobby view
- Fix issue with the constraints and examples field in `ModifyQuestionModal`

## Screenshots (if applicable)
<!-- [Include screenshots or images to visually showcase the changes, if applicable.] -->
Matched view:
<img width="308" alt="image" src="https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g05/assets/7760148/02f4e59b-1b8c-4d52-b674-3b888085e4af">

Question selection:
<img width="306" alt="image" src="https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g05/assets/7760148/7a1c3f3b-eca5-478b-a24e-8136dc801675">

## Checklist
- [x] I have checked that the changes included in the PR are intended to merge to `master` or any destination branch.
- [x] I have verified that the new changes do not break any existing functionalities, unless the new changes are intended and have approved by the team.
- [x] I will take care of the merging and delete the side-branch after the PR is merged.

## Additional Notes/References
- Question wrapper did not implement `throwAndLogError` as server action are unable to throw custom exceptions to client component.

tryyang2001 commented on frontend/package.json:
```diff
@@ -10,6 +10,12 @@
+    "setup:question-service": "cd ../backend/question-service && npm i",
+    "setup:collaboration-service": "cd ../backend/collaboration-service && npm i",
+    "setup:matching-service": "cd ../backend/matching-service && npm i",
+    "setup-all": "concurrently \"npm run setup:frontend\" \"npm run setup:user-service\" \"npm run setup:question-service\" \"npm run setup:matching-service\" \"npm run setup:collaboration-service\"",
```
Thanks for creating this tool!!! Was about to create a PR to do that xD
@tlyi @xingjie99 @weidak for your reference.

weidak replied:
👍🏻

tryyang2001 commented:
Just an update of the bug that I reported yesterday. This is regarding the matching modal does not appear. Here is how to reproduce the bug:
1. For two users, try to match in the first round. Any topic is fine (I believe). For my case, I used (Python, Easy, Array). 
2. Once both parties accept the match and has been redirected to the collaboration workspace, exit the room by either clicking the end session button or click the logo to back to home page. 
3. Try clicking the `Get Matched` button again, and the modal will not appear.

Possible guess: 
I think it might be related to the need of doing page refresh. Because every time when I encountered this, refresh the page and it will back to normal.

tryyang2001 commented:
Another flaw that I found is that even though the two parties have clicked the ready button, if one person refreshed his page, then he gets disconnected. Not sure if this is the intended behaviour? One issue might be that the room host click the start peerprep button but then like 0.001s before the button click the user refreshes his page... ?
![image](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g05/assets/69421979/656950da-a0ac-4023-a04c-b6e3f566b368)
After `John Doe` refresh his page:
![image](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g05/assets/69421979/ad351aa9-3460-4519-9eae-76b88fa7834a)

Ah and another potential issue from the user experience POV is that if both parties have clicked the ready button, but the room host does not click the `start peerprep` button and then the partner can only wait haha. Ok even though for now he can simply quit the session by page refreshing. So if this page refresh is not intended and then after you fixed that issue, you might also want to consider setting timeout/auto-start for this issue.

tryyang2001 commented:
Ui issue again:
As for now if the topic tags are too many, then the overflowed topic cannot be seen. Do you think if it is possible to allow horizontal scrolling in your dropdown list?
![image](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g05/assets/69421979/4625c81c-ae55-49fa-a5f2-1e16eca3b17e)

tryyang2001 commented:
Another bug that I found is that when this thing comes out:
![image](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g05/assets/69421979/63339c4e-83c5-444e-85da-435a1400b834)
Entering the collaboration page from the room host side will have this error, but this error does not break any functionality:
![image](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g05/assets/69421979/5b213de2-a976-424f-aeb9-7b9522559bb6)

If you want to reproduce this error:
1. For two users A and B, set A preferences as (Python, Medium, Array, String), and B preferences as (Python, Medium, String). Let's us assume that A is the room host.
2. Accept the match, pick any of the question in the dropdown list, and enter the collaboration workspace.
3. You will see the error toast occurs at A, this only occurs at the host side.

So I have a follow-up question regarding this issue:
I was actually thinking that can we do random question assignment instead of letting the room host pick a question? So that when we have the history service, we can remove questions that are already attempted by either one of the room participants, or at least reduce the likelihood of getting attempted questions.

c0j0s commented:
Thankss @tryyang2001 for finding the bugs.

> Just an update of the bug that I reported yesterday. This is regarding the matching modal does not appear. Here is how to reproduce the bug:

Fixed

> Another flaw that I found is that even though the two parties have clicked the ready button, if one person refreshed his page, then he gets disconnected. Not sure if this is the intended behaviour? One issue might be that the room host click the start peerprep button but then like 0.001s before the button click the user refreshes his page... ?

It is the intended behavior. Once the page got refreshed, the socket gets disconnected, even if the user request matching again, the socket Id is different already, so there's no way to rejoin the user unless matching service have knowledge of the user database. so I guess the simplest approach is to disconnect them.

> Ui issue again:

added a tooptip for it.

> I was actually thinking that can we do random question assignment instead of letting the room host pick a question? So that when we have the history service, we can remove questions that are already attempted by either one of the room participants, or at least reduce the likelihood of getting attempted questions.

no idea what cause the runtime error. as for random assignment I dont think it matters, I can filter the options when history service is ready. but that raises another question is that who's history should we take into account.

> server action are unable to throw custom exceptions to client component.

let's continue this discussion over #47

tryyang2001 reviewed (approved):
LGTM, the matching now looks very cool!!
